examples/vision/simsiam.py

```diff
@@ -0,0 +1,434 @@
+    # Affine transformations can disturb the natural range of
+    # RGB images, hence this is needed.
+    x = tf.clip_by_value(x, 0, 255)
+    return x
+
+
+def color_drop(x):
+    x = tf.image.rgb_to_grayscale(x)
+    x = tf.tile(x, [1, 1, 3])
+    return x
+
+
+# This function is referred from here:
+# https://github.com/google-research/simclr/blob/master/data_util.py.
+def random_apply(func, x, p):
+    return tf.cond(
+        tf.less(
+            tf.random.uniform([], minval=0, maxval=1, dtype=tf.float32),
+            tf.cast(p, tf.float32),
+        ),
+        lambda: func(x),
+        lambda: x,
+    )
+
+
+def custom_augment(image):
+    # As discussed in the SimCLR paper, the series of augmentation
+    # transformations (except for random crops) need to be applied
+    # randomly to impose translational invariance.
+    image = flip_random_crop(image)
+    image = random_apply(color_jitter, image, p=0.8)
+    image = random_apply(color_drop, image, p=0.2)
+    return image
+
+
+"""
+It should be noted that an augmentation pipeline is generally dependent on various
+properties of the dataset we are dealing with. For example, if images in the dataset are
+heavily object-centric then taking random crops with a very high probability may hurt the
+training performance.
+
+Let's now apply our augmentation pipeline to our dataset and visualize a few outputs. 
+"""
+
+"""
+## Convert the data into TensorFlow `Dataset` objects
+
+Here we create two different versions of our dataset *without* any ground-truth labels. 
+"""
+
+ssl_ds_one = tf.data.Dataset.from_tensor_slices(x_train)
+ssl_ds_one = (
+    ssl_ds_one.shuffle(1024, seed=SEED)
+    .map(custom_augment, num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+
+ssl_ds_two = tf.data.Dataset.from_tensor_slices(x_train)
+ssl_ds_two = (
+    ssl_ds_two.shuffle(1024, seed=SEED)
+    .map(custom_augment, num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+
+# We then zip both of these datasets.
+ssl_ds = tf.data.Dataset.zip((ssl_ds_one, ssl_ds_two))
+
+# Visualize a few augmented images.
+sample_images_one = next(iter(ssl_ds_one))
+plt.figure(figsize=(10, 10))
+for n in range(25):
+    ax = plt.subplot(5, 5, n + 1)
+    plt.imshow(sample_images_one[n].numpy().astype("int"))
+    plt.axis("off")
+plt.show()
+
+# Ensure that the different versions of the dataset actually contain
+# identical images.
+sample_images_two = next(iter(ssl_ds_two))
+plt.figure(figsize=(10, 10))
+for n in range(25):
+    ax = plt.subplot(5, 5, n + 1)
+    plt.imshow(sample_images_two[n].numpy().astype("int"))
+    plt.axis("off")
+plt.show()
+
+"""
+Notice that the images in `samples_images_one` and `sample_images_two` are essentially
+the same but are augmented differently.
+"""
+
+"""
+## Defining the encoder and the predictor
+
+We use an implementation of ResNet20 that is specifically configured for the CIFAR10
+dataset. The code is taken from the
```

Comment:
What are the benefits of using this model here rather than e.g. `applications.ResNet50`?

I see that the model uses a 32x32 input -- usually that's an antipattern for small image processing. It usually works better to upscale the images first.

https://github.com/GoogleCloudPlatform/keras-idiomatic-programmer/blob/master/zoo/resnet/resnet_cifar10_v2.py#L159

Reply:
I had tried with ResNet50. Because of its capacity, it easily leads to representation collapse. This is also studied in all the works I have mentioned throughout this example.